### PR TITLE
Implementation of AuthorizationCache

### PR DESCRIPTION
**Pull request by Pietfried**
- Added handling for ClearCache.req
- Improved updating and deleting of cache entries

### CODE REVIEW

File: lib/ocpp/v201/database_handler.cpp [hunk 1]
```diff
@@ -105,6 +114,33 @@ std::optional<IdTokenInfo> DatabaseHandler::get_auth_cache_entry(const std::stri
...
+        std::string sql = "DELETE FROM AUTH_CACHE WHERE ID_TOKEN_HASH = @id_token_hash";
+        sqlite3_stmt* stmt;
+        if (sqlite3_prepare_v2(this->db, sql.c_str(), sql.size(), &stmt, NULL) != SQLITE_OK) {
+            EVLOG_error << "Could not prepare insert statement: " << sqlite3_errmsg(this->db);
+            return;
+        }
+
+        sqlite3_bind_text(stmt, 1, id_token_hash.c_str(), id_token_hash.length(), NULL);
+        if (sqlite3_step(stmt) != SQLITE_DONE) {
+            EVLOG_error << "Could not delete from table: " << sqlite3_errmsg(db);
```
Comment (maaikez) [inline lib/ocpp/v201/database_handler.cpp:128]: Minor: here you just use db and below and above this->db

File: lib/ocpp/v201/database_handler.cpp [hunk 2]
```diff
@@ -105,6 +114,33 @@ std::optional<IdTokenInfo> DatabaseHandler::get_auth_cache_entry(const std::stri
...
+        }
+        if (sqlite3_finalize(stmt) != SQLITE_OK) {
+            EVLOG_error << "Error deleting from table: " << sqlite3_errmsg(this->db);
+        }
+    } catch (const json::exception& e) {
```
Comment (maaikez) [inline lib/ocpp/v201/database_handler.cpp:133]: Curious what json has to do with this?

File: lib/ocpp/v201/charge_point.cpp [hunk 3]
```diff
@@ -814,20 +828,41 @@ void ChargePoint::handle_reset_req(Call<ResetRequest> call) {
...
-                // TODO(piet): E05.FR.03
-                // Energy delivery to the EV SHALL be allowed until the amount of energy specified in
-                // MaxEnergyOnInvalidId has been reached.
+    if (msg.idTokenInfo.has_value()) {
+        // C10.FR.05
+        if (this->device_model->get_optional_value<bool>(ControllerComponentVariables::AuthCacheCtrlrEnabled)
+                .value_or(true)) {
+            this->database_handler->insert_auth_cache_entry(utils::sha256(id_token), msg.idTokenInfo.value());
+        }
+        if (msg.idTokenInfo.value().status != AuthorizationStatusEnum::Accepted) {
```
Comment (maaikez) [inline lib/ocpp/v201/charge_point.cpp:854]: Missing C10.FR.03 here, maybe add a todo? Or is it somewhere else?

File: lib/ocpp/v201/charge_point.cpp [hunk 4]
```diff
@@ -188,28 +186,35 @@ AuthorizeResponse ChargePoint::validate_token(const IdToken id_token, const std:
...
+            const auto cache_entry = this->database_handler->get_auth_cache_entry(hashed_id_token);
+            if (cache_entry.has_value()) {
+                if ((cache_entry.value().cacheExpiryDateTime.has_value() and
+                     cache_entry.value().cacheExpiryDateTime.value().to_time_point() < DateTime().to_time_point())) {
+                    this->database_handler->delete_auth_cache_entry(hashed_id_token);
+                } else if (this->device_model->get_value<bool>(ControllerComponentVariables::LocalPreAuthorize) and
+                           cache_entry.value().status == AuthorizationStatusEnum::Accepted) {
+                    response.idTokenInfo = cache_entry.value();
+                    return response;
+                }
```
Comment (a-w50) [inline lib/ocpp/v201/charge_point.cpp:199]: Like in the device_model review, you can get this nested `if-thing` into a linear flow

File: lib/ocpp/v201/database_handler.cpp (hunk 2 above)
Comment (Pietfried) [reply lib/ocpp/v201/database_handler.cpp:133]: copy mistake

File: lib/ocpp/v201/charge_point.cpp (hunk 3 above)
Comment (Pietfried) [reply lib/ocpp/v201/charge_point.cpp:868]: C10.FR.03 is covered by line L216 .

**Review by maaikez — APPROVED** [review]

**Pull request merged**